File: check_barcodes.mjs
// Quick check: how many tickets have a valid codeBarres in Firebase
import { initializeApp } from 'firebase/app';
import { getDatabase, ref, get } from 'firebase/database';

const firebaseConfig = {
  apiKey: "AIzaSyA1GwY_Cn-w2jHvLvwS7-nbDQYnEDkyJTA",
  databaseURL: "https://fest-rev-default-rtdb.firebaseio.com",
  projectId: "fest-rev",
};

const app = initializeApp(firebaseConfig);
const db = getDatabase(app);

async function checkBarcodes() {
  console.log("Fetching all tickets from Firebase...");
  const snapshot = await get(ref(db, 'tickets'));
  const data = snapshot.val();

  if (!data) {
    console.log("No tickets found.");
    process.exit();
  }

  const entries = Object.entries(data);
  const total = entries.length;
  let hasBarcode = 0;
  let emptyBarcode = 0;
  let zeroBarcode = 0;
  const sampleWith = [];
  const sampleWithout = [];

  for (const [ticketId, ticket] of entries) {
    let cb = ticket.codeBarres;
    // Also check raw Excel column variants
    if (cb === undefined || cb === null) {
      cb = ticket["Code-barres"] || ticket["code-barres"] || ticket["Code barres"] || ticket["Code_barres"] || null;
    }

    const cbStr = String(cb ?? "").trim();
    const isEmpty = !cb || cbStr === "" || cbStr === "0";

    if (isEmpty) {
      emptyBarcode++;
      if (sampleWithout.length < 5) {
        sampleWithout.push({
          id: ticketId,
          nCommande: ticket.nCommande || ticket["N_ de commande"] || "?",
          nBillet: ticket.nBillet || ticket["N_ billet"] || "?",
          codeBarres: cb,
          allKeys: Object.keys(ticket).slice(0, 25),
        });
      }
    } else {
      hasBarcode++;
      if (sampleWith.length < 5) {
        sampleWith.push({
          id: ticketId,
          nCommande: ticket.nCommande || ticket["N_ de commande"] || "?",
          nBillet: ticket.nBillet || ticket["N_ billet"] || "?",
          codeBarres: cb,
        });
      }
    }

    if (cbStr === "0") zeroBarcode++;
  }

  console.log("\n" + "=".repeat(60));
  console.log(`TOTAL TICKETS:        ${total}`);
  console.log(`WITH codeBarres:      ${hasBarcode}  (${(hasBarcode/total*100).toFixed(1)}%)`);
  console.log(`WITHOUT codeBarres:   ${emptyBarcode}  (${(emptyBarcode/total*100).toFixed(1)}%)`);
  console.log(`  (of which value=0): ${zeroBarcode}`);
  console.log("=".repeat(60));

  if (sampleWith.length > 0) {
    console.log(`\n--- Sample WITH barcode (${sampleWith.length}) ---`);
    for (const s of sampleWith) {
      console.log(`  nCommande=${s.nCommande}, nBillet=${s.nBillet}, codeBarres=${s.codeBarres}`);
    }
  }

  if (sampleWithout.length > 0) {
    console.log(`\n--- Sample WITHOUT barcode (${sampleWithout.length}) ---`);
    for (const s of sampleWithout) {
      console.log(`  nCommande=${s.nCommande}, nBillet=${s.nBillet}, codeBarres=${s.codeBarres}`);
      console.log(`    Fields: ${s.allKeys.join(", ")}`);
    }
  }

  // Show all unique field names from first 50 tickets
  const allKeys = new Set();
  let count = 0;
  for (const [, ticket] of entries) {
    if (count >= 50) break;
    Object.keys(ticket).forEach(k => allKeys.add(k));
    count++;
  }
  console.log(`\n--- All unique field names (from first 50 tickets) ---`);
  const sorted = [...allKeys].sort();
  console.log(sorted.join("\n  "));

  process.exit(0);
}

checkBarcodes().catch(err => {
  console.error("Error:", err);
  process.exit(1);
});

File: check_barcodes.py
#!/usr/bin/env python3
"""Quick check: how many tickets have a valid codeBarres in Firebase."""
import json, urllib.request

url = "https://fest-rev-default-rtdb.firebaseio.com/tickets.json"
print("Fetching all tickets from Firebase...")
with urllib.request.urlopen(url) as resp:
    data = json.loads(resp.read())

if not data:
    print("No tickets found.")
    exit()

total = len(data)
has_barcode = 0
empty_barcode = 0
zero_barcode = 0
sample_with = []
sample_without = []

for ticket_id, ticket in data.items():
    cb = ticket.get("codeBarres", None)
    # Also check common raw Excel column variants
    if cb is None:
        cb = ticket.get("Code-barres", ticket.get("code-barres", ticket.get("Code barres", ticket.get("code barres", None))))

    if cb is None or cb == "" or cb == 0 or cb == "0":
        empty_barcode += 1
        if len(sample_without) < 3:
            sample_without.append({
                "id": ticket_id,
                "nCommande": ticket.get("nCommande", ticket.get("N_ de commande", "?")),
                "nBillet": ticket.get("nBillet", ticket.get("N_ billet", "?")),
                "codeBarres": cb,
                "all_keys": list(ticket.keys())[:20]
            })
    else:
        has_barcode += 1
        if len(sample_with) < 3:
            sample_with.append({
                "id": ticket_id,
                "nCommande": ticket.get("nCommande", ticket.get("N_ de commande", "?")),
                "nBillet": ticket.get("nBillet", ticket.get("N_ billet", "?")),
                "codeBarres": cb
            })

    if str(cb) == "0":
        zero_barcode += 1

print(f"\n{'='*60}")
print(f"TOTAL TICKETS:        {total}")
print(f"WITH codeBarres:      {has_barcode}  ({has_barcode/total*100:.1f}%)")
print(f"WITHOUT codeBarres:   {empty_barcode}  ({empty_barcode/total*100:.1f}%)")
print(f"  (of which value=0): {zero_barcode}")
print(f"{'='*60}")

if sample_with:
    print(f"\n--- Sample WITH barcode ({len(sample_with)}) ---")
    for s in sample_with:
        print(f"  nCommande={s['nCommande']}, nBillet={s['nBillet']}, codeBarres={s['codeBarres']}")

if sample_without:
    print(f"\n--- Sample WITHOUT barcode ({len(sample_without)}) ---")
    for s in sample_without:
        print(f"  nCommande={s['nCommande']}, nBillet={s['nBillet']}, codeBarres={s['codeBarres']}")
        print(f"    Fields in this ticket: {s['all_keys']}")

# Check what field names exist that might contain barcodes
print(f"\n--- All unique field names across first 50 tickets ---")
all_keys = set()
for i, (tid, ticket) in enumerate(data.items()):
    if i >= 50:
        break
    all_keys.update(ticket.keys())
for k in sorted(all_keys):
    print(f"  {k}")
